Kafka Validations

bilalba opened the pull request:
Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-events/blob/master/USERS.md).

<!--

Please leave your PR in draft if you don't need a review yet. 

To fix failing `CI / Codegen`, run `make codegen`. 

-->

bilalba committed:
webhook validations

Signed-off-by: Bilal Bakht Ahmad <tringingly@gmail.com>

bilalba committed:
add Exotic Kafka installer

dfarr commented on controllers/eventbus/validate.go:
```diff
@@ -8,8 +8,8 @@ import (
-	if eb.Spec.NATS == nil && eb.Spec.JetStream == nil {
-		return fmt.Errorf("invalid spec: either \"nats\" or \"jetstream\" needs to be specified")
+	if eb.Spec.NATS == nil && eb.Spec.JetStream == nil && eb.Spec.Kafka == nil {
+		return fmt.Errorf("invalid spec: either \"nats\", \"jetstream\" or \"kafka\" needs to be specified")
```
```suggestion
		return fmt.Errorf("invalid spec: either \"nats\", \"jetstream\", or \"kafka\" needs to be specified")
```

dfarr commented on controllers/eventbus/validate.go:
```diff
@@ -36,5 +36,14 @@ func ValidateEventBus(eb *v1alpha1.EventBus) error {
+			return fmt.Errorf("\"exotic\" must be defined")
+		}
+		e := x.Exotic
+		if e.URL == "" {
```
```suggestion
		if x.Exotic.URL == "" {
```

One less variable to define

dfarr reviewed (changes requested):
This is awesome!

We need to add `kafka` [here](https://github.com/intuit-data-os/argo-events/blob/feature/kafka-eventbus/.github/workflows/ci.yaml#L125-L127) but let's hold off on doing that so that we can merge this PR with all the checks green. Then we can just push that change directly to the `feature/kafka-eventbus` branch.

dfarr committed:
Change order of EventBus reconcilation

Calling client.Status().Update() before calling client.Update()
persists the status for exotic evnetbuses, otherwise the status
is not perisisted.

Signed-off-by: David Farr <david_farr@intuit.com>

bilalba committed:
requeue reconciler

Signed-off-by: Bilal Bakht Ahmad <tringingly@gmail.com>

dfarr committed:
Make additional copy of obj to update eb status

Signed-off-by: David Farr <david_farr@intuit.com>

bilalba committed:
oxford comma

Co-authored-by: David Farr <david_farr@intuit.com>

bilalba committed:
stylistic change

Signed-off-by: Bilal Bakht Ahmad <tringingly@gmail.com>

dfarr reviewed (approved):
Nice!
